### DISCUSSION TITLE
Adding more built-in casters

### DISCUSSION THREAD

**Discussion by mauriciovillam** [spatie/laravel-data#6] [Ideas]
Hello,

I'm currently moving my project from spatie/data-transfer-object to spatie/laravel-data. I noticed that this project includes a cast by default (DateTimeInterface), wouldn't it be a good idea to implement spatie/data-transfer-object's [ArrayCaster](https://github.com/spatie/data-transfer-object/blob/master/src/Casters/ArrayCaster.php) and the equivalent of [DataTransferObjectCaster](https://github.com/spatie/data-transfer-object/blob/master/src/Casters/DataTransferObjectCaster.php)?

Thanks.

**Comment by rubenvanassche** [comment]
Hi @mauriciovillam,

In case of the `DataTransferObjectCaster` this automatically already happens, if you nest a Data object into another and provide an array with the correct data for that nested data object it will be automatically created.  (You can read more about it here: https://spatie.be/docs/laravel-data/v1/as-a-data-transfer-object/nesting)

For the `ArrayCaster`, the same thing happens when you have a `DataCollection` annotated with the data class, then an array of arrays that could be casted into data objects will be automatically casted. (More information here: https://spatie.be/docs/laravel-data/v1/as-a-data-transfer-object/collections)

**Comment by mauriciovillam** [comment]
@rubenvanassche 

Looks good, however, one of the use cases I had for ArrayCaster is setting the default value to an empty array `[]` to optional promoted properties in the constructor. I'm not sure if this is within the scope of the package but currently I can't do this without adding some boilerplate. This means, I can't do this (I know it's obvious why I can't do it but just to show what I'm trying to convey):

```
public function __construct(
        /** @var Step[] */
        public DataCollection $steps = []
) { }
```

**Edit nov. 8:** This issue may be resolved in PHP 8.1 due to being able to use new in initializers. (https://stitcher.io/blog/php-81-new-in-initializers)

On the other hand, DataCollection with annotated Data models in the *same* namespace are not resolved correctly. I'm getting this error:

```
Data collection property `steps` in `Data\Models\Plan` has an annotation that isn't a data object
```

`$steps` is instantiated like this in the constructor:

```
/** @var Step[] */
public DataCollection $steps,
```

If I change it to this it works:
```
/** @var \Data\Models\Step[] */
public DataCollection $steps,
```

The problem seems to occur in https://github.com/spatie/laravel-data/tree/main/src/Support/DataProperty.php:151. I see that this problem has already been foreseen with a TODO comment. We are probably re-inventing the wheel there and need to delegate the class' namespace resolver logic to an utility class/library. You can see a similar discussion here: https://github.com/php-annotations/php-annotations/issues/15.

**Comment by rubenvanassche** [comment]
Hi @mauriciovillam,

The problem should be solved by using version 1.1.0 of the package, it now has built in support for classes without FQN's 🎉 

As for the initializer problem, yeah PHP 8.1 is only two weeks away from us, I think that would be the best solution to tackle this issue.
